src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2.java

```diff
@@ -0,0 +1,283 @@
+            CountingReadFilter filterToApply = BaseRecalibratorDataflow2.readFilter(readsHeader);
+
+            bunny.stepEnd("set up bam input");
+
+            // Load the Variants and the Reference
+            //final VariantsDataflowSource variantsDataflowSource = new VariantsDataflowSource(baseRecalibrationKnownVariants, pipeline);
+            final ReferenceDataflowSource referenceDataflowSource = new ReferenceDataflowSource(pipeline.getOptions(), referenceURL, BQSR_REFERENCE_WINDOW_FUNCTION);
+            bunny.stepEnd("create referenceDataflowSource");
+            final SAMSequenceDictionary refDictionary = referenceDataflowSource.getReferenceSequenceDictionary(readsDictionary);
+            bunny.stepEnd("load ref sequence dictionary");
+
+            checkSequenceDictionaries(refDictionary, readsDictionary);
+            bunny.stepEnd("checkSequenceDictionaries");
+            PCollectionView<SAMSequenceDictionary> refDictionaryView = pipeline.apply(Create.of(refDictionary)).setName("refDictionary").apply(View.asSingleton());
+            bunny.stepEnd("create ref dictionary view");
+
+            // the granularity at which we'll want to assign work and read inputs.
+            final int bigShardSize = 1_000_000;
+            // the granularity at which we want to batch reads,variants,and reference for processing.
+            final int smallShardSize = 5_000;
+
+            List<SimpleInterval> shardedIntervals = ShardedReadsDatasource.cutToShards(intervals, bigShardSize);
+            // since we currently can only read variants at the client, that's the right place to populate the shards.
+            List<Variant> variants = VariantsDataflowSource.getVariantsList(baseRecalibrationKnownVariants);
+
+            bunny.stepEnd("load variants");
+            int margin = 1000;
+            ArrayList<AddContextDataToReadOptimized.ContextShard> shards = fillVariants(shardedIntervals, variants, margin);
+            bunny.stepEnd("sharding variants");
+
+            logger.info("Shipping "+shards.size()+" big shards.");
+
+            PCollection<AddContextDataToReadOptimized.ContextShard> shardsPCol = pipeline.apply(Create.of(shards));
+
+            PCollection<AddContextDataToReadOptimized.ContextShard> shardsWithContext = shardsPCol
+                // big shards of variants -> smaller shards with variants, reads. We take the opportunity to filter the reads as close to the source as possible.
+                .apply(ParDo.named("subdivideAndFillReads").of(AddContextDataToReadOptimized.subdivideAndFillReads(bam, smallShardSize, margin, filterToApply)))
+                // add ref bases to the shards
+                .apply(ParDo.named("fillContext").of(AddContextDataToReadOptimized.fillContext(referenceDataflowSource, refDictionaryView, offlineAuth))
+                    .withSideInputs(refDictionaryView));
```

Comment:
I can do it for you, sure. It'll be an awkwardly large function, though, even more so because it mixes client code and worker code. 
And it muddles the AddContextDataToReadOpt class which used to be purely worker code, to also contain client code.